torchani/data/new.py

```diff
@@ -0,0 +1,458 @@
+            species = np.array([species_dict[x] for x in molecule['species']])
+            self.data_species += list(np.tile(species, (num_conformations, 1)))
+            # energies
+            self.data_energies += list(molecule['energies'].reshape((-1, 1)).astype(np.float32))
+            if subtract_self_energies:
+                self_energies = np.array(sum([self_energies_dict[x] for x in molecule['species']]))
+                self.data_self_energies += list(np.tile(self_energies, (num_conformations, 1)).astype(np.float32))
+            # other properties
+            if enable_pkbar:
+                pbar.update(i)
+
+        if subtract_self_energies:
+            self.data_energies = np.array(self.data_energies) - np.array(self.data_self_energies)
+            import gc
+            del self.data_self_energies
+            del self_energies
+            gc.collect()
+
+        self.batch_size = batch_size
+        self.length = (len(self.data_species) + self.batch_size - 1) // self.batch_size
+        self.device = device
+
+        self.shuffled_index = np.arange(len(self.data_species))
+        np.random.shuffle(self.shuffled_index)
+
+        self.chunk_threshold = chunk_threshold
+        if not self.chunk_threshold:
+            self.chunk_threshold = np.inf
+
+        anidata.cleanup()
+        import gc
+        del num_conformations
+        del species
+        del anidata
+        gc.collect()
+
+    @functools.lru_cache(maxsize=None)
+    def __getitem__(self, index):
+
+        if index >= self.length:
+            raise IndexError()
+
+        batch_indices = slice(index * self.batch_size, (index + 1) * self.batch_size)
+        batch_indices_shuffled = self.shuffled_index[batch_indices]
+
+        batch_species = [self.data_species[i] for i in batch_indices_shuffled]
+        batch_coordinates = [self.data_coordinates[i] for i in batch_indices_shuffled]
+        batch_energies = [self.data_energies[i] for i in batch_indices_shuffled]
+
+        datas = [batch_species, batch_coordinates, batch_energies]
+
+        # get sort index
+        num_atoms_each_mole = [b.shape[0] for b in batch_species]
+        atoms = torch.tensor(num_atoms_each_mole, dtype=torch.int32)
+        sorted_atoms, sorted_atoms_idx = torch.sort(atoms, descending=False)
+
+        # sort each batch of data
+        for i, d in enumerate(datas):
+            datas[i] = self.sort_list_with_index(d, sorted_atoms_idx.numpy())
+
+        # get chunk size
+        output, count = torch.unique(atoms, sorted=True, return_counts=True)
+        counts = torch.cat((output.unsqueeze(-1).int(), count.unsqueeze(-1).int()), dim=-1)
+        chunk_size_list, chunk_max_list = split_to_chunks(counts, chunk_threshold=self.chunk_threshold * self.batch_size * 20)
+        chunk_size_list = torch.stack(chunk_size_list).flatten()
+
+        # split into chunks
+        datas[0] = self.split_list_with_size(datas[0], chunk_size_list.numpy())
```

Comment:
`data[0]` is hard to read, use its real name:
batch_species for 0, batch_coordinates for 1, and batch_energies for 2.